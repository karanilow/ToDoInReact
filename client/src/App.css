.container {
  box-sizing: border-box;
  padding: 25px;
  width: 100%;
  height: 800px;
}
.container > .row {
  box-sizing: border-box;
  display: flex;
  flex-flow: row wrap;
  align-items: center;
  justify-content: left;
  padding-left: 150px;
}
.controls {
  box-sizing: border-box;
  flex: 30%;
  padding: 20px;
}
h1 {
  text-align: center;
}
#delete,
#finish {
  background-color: rgb(241, 241, 241);
  color: black;
  padding: 20px;
  text-align: center;
  border-radius: 50%;
  border: 1px solid rgb(228, 84, 84);
  display: block;
  font-size: 20px;
  box-shadow: 0 5px 10px 0 rgb(0, 0, 0, 0.2);
  cursor: pointer;
  width: 75px;
  margin: 13px;
}
#delete:hover {
  background-color: rgb(228, 84, 84);
}
#delete:active,
#finish:active {
  border: 1px solid rgb(241, 241, 241);
}
#finish {
  border: 1px solid rgb(137, 238, 123);
}
#finish:hover {
  background-color: rgb(137, 238, 123);
}

.tasks {
  box-sizing: border-box;
  flex: 70%;
  padding: 20px;
  overflow: auto;
  max-width: 700px;
}

.tasks h2 {
  font-weight: 300;
}
.singleTask {
  display: flex;
}

.singleTaskName {
  padding: 10px 20px;
  margin-bottom: 10px;
  display: flex;
  border: 0.08em solid silver;
  border-radius: 20px;
}

.default {
  background: whitesmoke;
  margin: 0.2em;
}
.finished {
  background-color: rgb(137, 238, 123);
  margin: 0.2em;
}
.isSelected {
  border-color: red;
}
.isNotSelected {
  border-color: silver;
}

.time {
  font-size: 10px;
  padding-left: 5px;
  padding-right: 5px;
  color: #999;
}

.form {
  box-sizing: border-box;
  flex: 100%;
  padding: 20px;
  width: 100%;
  min-width: 300px;
  max-width: 700px;
}

.form h2 {
  font-weight: 300;
}

.form form {
  align-items: baseline;
}

.form input {
  height: 30px;
  padding-left: 10px;
  border: 2px solid #c1c1c1;
  margin-right: 10px;
  border-radius: 20px;
  flex: 3;
  box-sizing: border-box;
}

.form input:first-child {
  flex: 1;
}

.form input:focus {
  border: 2px solid #385997;
}

.form button {
  color: #999;
  border-radius: 5px;
  border: 1px solid #999;
}

.form button:hover {
  cursor: pointer;
  color: #385997;
  border-color: #385997;
}
