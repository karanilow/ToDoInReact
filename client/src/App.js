import React, { Component } from "react";
import "./App.css";
import TaskInput from "./TaskInput";
import TaskBox from "./TaskBox";

class App extends Component {
  constructor() {
    super();
    this.home = {
      _id: "root",
      name: "home",
      partOfGroup: "none",
    };
    this.pollInterval = null;
    this.state = {
      error: null,
      author: "",
      taskName: "",
      taskList: [],
      selectedTasksId: [],
      currentGroup: this.home,
    };
  }

  componentDidMount() {
    this.loadCommentsFromServer();
    if (!this.pollInterval) {
      this.pollInterval = setInterval(this.loadCommentsFromServer, 5000);
    }
  }

  componentWillUnmount() {
    if (this.pollInterval) clearInterval(this.pollInterval);
    this.pollInterval = null;
  }

  loadCommentsFromServer = () => {
    fetch("/api/tasks/")
      .then((data) => data.json())
      .then((res) => {
        if (!res.success) this.setState({ error: res.error });
        else this.setState({ taskList: res.data });
      });
  };

  // Arrow fx for binding
  handleTaskClick = (id) => {
    const { selectedTasksId } = this.state;
    const isTaskSelected = selectedTasksId.includes(id);

    isTaskSelected
      ? selectedTasksId.splice(selectedTasksId.indexOf(id), 1)
      : selectedTasksId.splice(this.sortedIndex(selectedTasksId, id), 0, id);

    this.setState({ selectedTasksId });
  };

  // Arrow fx for binding
  handleIconClick = (id) => {
    const { taskList, selectedTasksId } = this.state;
    this.setState({
      currentGroup: taskList.find((task) => task._id === id),
    });
    selectedTasksId.length = 0;
    this.setState({ selectedTasksId });
  };

  // Arrow fx for binding
  handlePillsClick = (id) => {
    const { taskList, selectedTasksId } = this.state;
    id === "root"
      ? this.setState({ currentGroup: this.home })
      : this.setState({
          currentGroup: taskList.find((task) => task._id === id),
        });
    selectedTasksId.length = 0;
    this.setState({ selectedTasksId });
  };

  sortedIndex(array, value) {
    var low = 0,
      high = array.length;

    while (low < high) {
      var mid = (low + high) >>> 1;
      array[mid] < value ? (low = mid + 1) : (high = mid);
    }
    return low;
  }

  // arrow fx for binding
  isTaskSelected = (id) => {
    const { selectedTasksId } = this.state;

    return selectedTasksId.includes(id) ? "isSelected" : "isNotSelected";
  };

  // arrow fx for binding
  saveTaskEntry = (taskName, isTaskGroup, partOfGroup) => {
    fetch("/api/tasks", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ taskName, isTaskGroup, partOfGroup }),
    })
      .then((res) => res.json())
      .then((res) => {
        if (!res.success)
          this.setState({ error: res.error.message || res.error });
        else this.setState({ author: "", taskName: "", error: null });
      });
    this.loadCommentsFromServer();
  };

  updateTaskStatus(id, status) {
    fetch(`/api/tasks/${id}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ status }),
    })
      .then((res) => res.json())
      .then((res) => {
        if (!res.success)
          this.setState({ error: res.error.message || res.error });
      });
  }

  deleteTask(id) {
    fetch(`api/tasks/${id}`, {
      method: "DELETE",
      headers: { "Content-Type": "application/json" },
    })
      .then((res) => res.json())
      .then((res) => {
        if (!res.success)
          this.setState({ error: res.error.message || res.error });
      });
  }

  getRelativesIdFrom(id) {
    var { taskList } = this.state;
    var relativeChildrensId = [];
    const task = taskList.find((task) => task._id === id);

    if (task !== undefined) {
      var exTaskList = [];
      exTaskList = taskList.slice();
      exTaskList.splice(exTaskList.indexOf(task), 1);
      if (task.isTaskGroup) {
        var otherRelatives = exTaskList.filter((t) =>
          t.partOfGroups.includes(task._id)
        );
        while (otherRelatives.length > 0) {
          const relative = otherRelatives.pop();
          relativeChildrensId.push(relative._id);
          if (relative.isTaskGroup) {
            exTaskList.splice(exTaskList.indexOf(relative), 1);
            let newOtherRelatives = exTaskList.filter((t) =>
              t.partOfGroups.includes(relative._id)
            );
            otherRelatives = [...otherRelatives, ...newOtherRelatives];
          }
        }
      }
    } else {
      console.log("Error : This task doesn't exist !!");
    }
    return relativeChildrensId;
  }

  // Arrow fx for binding
  handleFinishedClick = () => {
    const { selectedTasksId, taskList } = this.state;
    const finishedTasks = [];
    const defaultTasks = [];

    taskList.forEach((task) => {
      if (selectedTasksId.indexOf(task._id) > -1) {
        if (task.status === "finished") {
          finishedTasks.push(task);
          task.status = "default";
        } else {
          defaultTasks.push(task);
          task.status = "finished";
        }
      }
    });

    finishedTasks.forEach((task) => this.updateTaskStatus(task._id, "default"));
    defaultTasks.forEach((task) => this.updateTaskStatus(task._id, "finished"));

    selectedTasksId.length = 0;

    this.setState({
      selectedTasksId,
      taskList,
    });
  };

  // arrow fx for binding
  handleRemoveClick = () => {
    const { taskList, selectedTasksId } = this.state;
    var newTaskList = taskList.slice();

    selectedTasksId.forEach((id) => {
      let i = newTaskList.findIndex((c) => c._id === id);
      let relatives = this.getRelativesIdFrom(id);
      newTaskList = [...newTaskList.slice(0, i), ...newTaskList.slice(i + 1)];
      console.log(id, relatives);
      this.deleteTask(id);
      relatives.forEach((id) => {
        this.deleteTask(id);
      });
    });

    selectedTasksId.length = 0;

    this.setState({
      taskList: newTaskList,
      selectedTasksId,
    });
  };

  // arrow fx for binding
  getGroupList = () => {
    const { taskList } = this.state;
    return taskList.filter((task) => task.isTaskGroup);
  };

  render() {
    return (
      <div className="container">
        <h1>React To Do's</h1>
        <div className="row">
          <div className="tasks">
            <h2>Tasks: </h2>
            <TaskBox
              handleTaskClick={this.handleTaskClick}
              handleIconClick={this.handleIconClick}
              handlePillsClick={this.handlePillsClick}
              isTaskSelected={this.isTaskSelected}
              data={this.state.taskList}
              currentGroup={this.state.currentGroup}
              home={this.home}
            />
          </div>
          <div className="controls">
            <button
              name="deleteButton"
              id="delete"
              type="button"
              onClick={() => this.handleRemoveClick()}
            >
              <i className="fas fa-trash"></i>
            </button>
            <button
              name="finishedButton"
              id="finish"
              type="button"
              onClick={() => this.handleFinishedClick()}
            >
              <i className="fas fa-check"></i>
            </button>
          </div>
        </div>
        <div className="row">
          <div className="form">
            <h2>Quick task : </h2>
            <TaskInput
              groupList={this.getGroupList}
              onStored={this.saveTaskEntry}
              currentGroup={this.state.currentGroup}
            />
          </div>
        </div>
        {this.state.error && <p>{this.state.error}</p>}
      </div>
    );
  }
}

export default App;
