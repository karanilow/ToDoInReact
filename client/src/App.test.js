import React from "react";
import { shallow } from "enzyme";
import { expect } from "chai";

import App from "./App";

describe("<App />", () => {
  it("renders without crashing", () => {
    const wrapper = shallow(<App />);
  });
  it("selected tasks should be removed correctly", () => {
    const wrapper = shallow(<App />);
    const initTaskList = Array.from({ length: 20 }, () =>
      Math.floor(Math.random() * 20).toString()
    );
    const initSelectedTasksIndices = [
      0,
      Math.floor(Math.random() * 16 + 1),
      19,
    ];
    const initFinishedTasksIndices = [
      0,
      initSelectedTasksIndices[1],
      initSelectedTasksIndices[1] + 1,
    ];
    const finalTaskList = initTaskList;
    finalTaskList.splice(initSelectedTasksIndices[1], 1);
    finalTaskList.shift();
    finalTaskList.pop();
    const finalFinishedTasksInidices = initFinishedTasksIndices;
    finalFinishedTasksInidices.pop();
    finalFinishedTasksInidices.pop();
    finalFinishedTasksInidices[0] = initFinishedTasksIndices[2] - 2;

    wrapper.setState({
      taskList: initTaskList,
      selectedTasksIndices: initSelectedTasksIndices,
      finishedTasksIndices: initFinishedTasksIndices,
    });

    wrapper.find('[name="deleteButton"]').simulate("click");

    expect(wrapper.state("taskList")).to.be.equal(finalTaskList);
    expect(wrapper.state("selectedTasksIndices")).to.be.empty();
    expect(wrapper.state("finishedTasksIndices")).to.be.equal(
      finalFinishedTasksInidices
    );
  });
  it("selected tasks should be marked as finished correctly", () => {
    const wrapper = shallow(<App />);
    const initTaskList = Array.from({ length: 20 }, () =>
      Math.floor(Math.random() * 20).toString()
    );
    const initSelectedTasksIndices = [
      0,
      Math.floor(Math.random() * 15 + 1),
      19,
    ];
    const initFinishedTasksIndices = [
      0,
      initSelectedTasksIndices[1],
      initSelectedTasksIndices[1] + 1,
    ];

    const finalFinishedTasksIndices = initFinishedTasksIndices;
    finalFinishedTasksIndices.pop();
    finalFinishedTasksIndices.pop();
    finalFinishedTasksIndices.splice(0, 0, 19);

    wrapper.setState({
      taskList: initTaskList,
      selectedTasksIndices: initSelectedTasksIndices,
      finishedTasksIndices: initFinishedTasksIndices,
    });

    wrapper.find('[name="finishedButton"]').simulate("click");

    expect(wrapper.state("taskList")).to.be.equal(initTaskList);
    expect(wrapper.state("selectedTasksIndices")).to.be.empty();
    expect(wrapper.state("finishedTasksIndices")).to.be.equal(
      finalFinishedTasksIndices
    );
  });
  it("should match its reference snapshot", () => {
    const wrapper = shallow(<App />);

    expect(wrapper).to.matchSnapshot();
  });
});
