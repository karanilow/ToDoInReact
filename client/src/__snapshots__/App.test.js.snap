// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match its reference snapshot 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <p>
      To do list in React !
    </p>
  </header>
  <button
    name="deleteButton"
    onClick={[Function]}
    type="button"
  >
    Delete
  </button>
  <button
    name="finishedButton"
    onClick={[Function]}
    type="button"
  >
    Finished
  </button>
  <TaskInput
    onStored={[Function]}
  />
</div>
`;
